src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -621,6 +621,11 @@ static int am_recv_event(struct fi_cq_tagged_entry *wc, MPIR_Request * rreq)
         mpi_errno = MPIDI_OFI_am_enqueue_unordered_msg(orig_buf);
         MPIR_ERR_CHECK(mpi_errno);
         goto fn_exit;
+    } else {
+        if (am_hdr->am_type == MPIDI_AMTYPE_RDMA_READ) {
+            orig_buf = MPL_malloc(wc->len, MPL_MEM_OTHER);
+            memcpy(orig_buf, wc->buf, wc->len);
+        }
     }
```

Comment:
Three comments for the alignment fix:
1. Why do you define the copy in the else branch? It seems unnecessary.

2. Does the issue happen only when `NEEDS_STRICT_ALIGNMENT` is on? If yes, can we guard it with ifdef macro similar to the am_hdr copy?

3. Since the "payload" is always a fixed size `MPIDI_OFI_lmt_msg_payload_t`, can we just define a static buffer rather than calling malloc? (similar to the `am_hdr` copy code)

Reply:
1. because if the packet is not in order, it would be cached. In `MPIDI_OFI_am_enqueue_unordered_msg`, the data will be aligned. So only when the packet is in order, we would need to align the data.
2. yes, we can guard it with `NEEDS_STRICT_ALIGNMENT`
3. no it is not possible, because we only know size of ofi hdr (`MPIDI_OFI_am_header_t`) and payload (`MPIDI_OFI_lmt_msg_payload_t`), but we don't know the payload header size (`am_hdr`). We must use malloc for this mem alignment

Reply:
> * because if the packet is not in order, it would be cached. In `MPIDI_OFI_am_enqueue_unordered_msg`, the data will be aligned. So only when the packet is in order, we would need to align the data.

When it is out-of-order, it always `goto fn_exit;`. That is why I think the else branch is unnecessary.

> * no it is not possible, because we only know size of ofi hdr (`MPIDI_OFI_am_header_t`) and payload (`MPIDI_OFI_lmt_msg_payload_t`), but we don't know the payload header size (`am_hdr`). We must use malloc for this mem alignment

I am confused. `am_hdr` is already aligned. The only address that might be unaligned is `p_data`. You will have trouble when accessing it as `MPIDI_OFI_lmt_msg_payload_t*` inside `MPIDI_OFI_handle_rdma_read`. Is it correct?

Reply:
- ok, I can delete `else`
- I got your point, I change it to only align `MPIDI_OFI_lmt_msg_payload_t`, please check

Reply:
The addr calculation for `orig_buf` and `p_data` seems redundant.

How about we do the alignment copy inside the `case` block after computed the p_data value? I.e.,
```
p_data = (char *) wc->buf + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
#if NEEDS_STRICT_ALIGNMENT
memcpy(amtype_rdma_payload, p_data, sizeof(MPIDI_OFI_lmt_msg_payload_t));
p_data = (void *) amtype_rdma_payload;
#endif
```

Reply:
you cannot use `wc->buf` since this packet may be the previous cached one; we have to use `orig_buf` here.
if that packet is the cached one, we don't need to call `memcpy` for memory alignment. so I don't recommend to put it in case block

Reply:
I see. Can you comment in the code that the copy is needed only for the first packet received from network?

Still, the extra addr calculation really bugs me. How about we change the code block to this:
```
#if NEEDS_STRICT_ALIGNMENT
p_data = amtype_rdma_payload;
#else
p_data = (char *) orig_buf + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
#endif

Reply:
sure, I can do it.